File: optur/search_space/distribution.py
from optur.errors import InCompatibleSearchSpaceError
from optur.proto.search_space_pb2 import Distribution, ParameterValue


def does_distribution_contain_value(distribution: Distribution, value: ParameterValue) -> bool:
    if distribution.HasField("int_distribution"):
        if not value.HasField("int_value"):
            return False
        return (
            distribution.int_distribution.low
            <= value.int_value
            <= distribution.int_distribution.high
        )
    if distribution.HasField("float_distribution"):
        if not value.HasField("double_value"):
            return False
        return (
            distribution.float_distribution.low
            <= value.double_value
            <= distribution.float_distribution.high
        )
    if distribution.HasField("categorical_distribution"):
        return any(value == v for v in distribution.categorical_distribution.choices)
    raise NotImplementedError("")


def extend_distribution(
    original_distribution: Distribution,
    new_distribution: Distribution,
) -> Distribution:
    if original_distribution.HasField("fixed_distribution"):
        if new_distribution.HasField("fixed_distribution"):
            values = [v for v in original_distribution.fixed_distribution.values]
            for new_v in new_distribution.fixed_distribution.values:
                if not any(new_v == v for v in values):
                    values.append(new_v)
            return Distribution(fixed_distribution=Distribution.FixedDistribution(values=values))
        if all(
            does_distribution_contain_value(new_distribution, value)
            for value in original_distribution.fixed_distribution.values
        ):
            return new_distribution
        raise InCompatibleSearchSpaceError("")
    if new_distribution.HasField("fixed_distribution"):
        if all(
            does_distribution_contain_value(original_distribution, value)
            for value in new_distribution.fixed_distribution.values
        ):
            return original_distribution
    if original_distribution == new_distribution:
        return original_distribution
    if original_distribution.HasField("categorical_distribution") and new_distribution.HasField(
        "categorical_distribution"
    ):
        values1 = original_distribution.categorical_distribution.choices
        values2 = new_distribution.categorical_distribution.choices
        if len(values1) == len(values2) and all(any(v1 == v2 for v2 in values2) for v1 in values1):
            return original_distribution
    raise InCompatibleSearchSpaceError("")  # TODO(tsuzuku)

File: src/search_space/distribution.rs
use crate::proto::optur;

pub fn int_distribution(low: i64, high: i64, log_scale: bool) -> optur::Distribution {
    optur::Distribution {
        distribution: Some(optur::distribution::Distribution::IntDistribution(
            optur::distribution::IntDistribution {
                low,
                high,
                log_scale,
            },
        )),
    }
}

pub fn float_distribution(low: f64, high: f64, log_scale: bool) -> optur::Distribution {
    optur::Distribution {
        distribution: Some(optur::distribution::Distribution::FloatDistribution(
            optur::distribution::FloatDistribution {
                low,
                high,
                log_scale,
            },
        )),
    }
}

pub fn categorical_distribution(choices: Vec<optur::ParameterValue>) -> optur::Distribution {
    optur::Distribution {
        distribution: Some(optur::distribution::Distribution::CategoricalDistribution(
            optur::distribution::CategoricalDistribution { choices },
        )),
    }
}

pub fn fixed_distribution(value: optur::ParameterValue) -> optur::Distribution {
    optur::Distribution {
        distribution: Some(optur::distribution::Distribution::FixedDistribution(
            optur::distribution::FixedDistribution { value: Some(value) },
        )),
    }
}

pub fn unknown_distribution(values: Vec<optur::ParameterValue>) -> optur::Distribution {
    optur::Distribution {
        distribution: Some(optur::distribution::Distribution::UnknownDistribution(
            optur::distribution::UnknownDistribution { values },
        )),
    }
}

pub fn contains(distribution: &optur::Distribution, parameter: &optur::ParameterValue) -> bool {
    match (
        &distribution.distribution.as_ref(),
        &parameter.value.as_ref(),
    ) {
        (Some(optur::distribution::Distribution::UnknownDistribution(dist)), _) => {
            panic!(); // Undefined.
        }
        (
            Some(optur::distribution::Distribution::IntDistribution(dist)),
            Some(optur::parameter_value::Value::IntValue(i)),
        ) => dist.low <= *i && *i <= dist.high,
        (
            Some(optur::distribution::Distribution::FloatDistribution(dist)),
            Some(optur::parameter_value::Value::DoubleValue(i)),
        ) => dist.low <= *i && *i <= dist.high,
        (Some(optur::distribution::Distribution::CategoricalDistribution(dist)), Some(p)) => {
            dist.choices.iter().any(|c| &c.value.as_ref().unwrap() == p)
        }
        (Some(optur::distribution::Distribution::FixedDistribution(dist)), Some(p)) => {
            &dist.value.as_ref().unwrap().value.as_ref().unwrap() == p
        }
        _ => false,
    }
}
